adapters/handlers/graphql/local/common_filters/authz.go

```diff
@@ -0,0 +1,71 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package common_filters
+
+import (
+	"fmt"
+
+	"github.com/weaviate/weaviate/entities/filters"
+	"github.com/weaviate/weaviate/entities/models"
+	"github.com/weaviate/weaviate/entities/search"
+	"github.com/weaviate/weaviate/usecases/auth/authorization"
+)
+
+func authorizePath(authorizer authorization.Authorizer, path *filters.Path, principal *models.Principal) error {
+	if path == nil {
+		return nil
+	}
+	if err := authorizer.Authorize(principal, authorization.READ, authorization.CollectionsData(path.Class.String())...); err != nil {
```

Comment:
Yes the `metadata` auth check has to go in the handler at the top so that introspection queries are forbidden in addition to protecting against information leakage issues due to the schema being stored in memory at that level